keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+
+This problem can be fixed if containerd, CRI and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index. This document details the changes needed to support pulling images per runtime class.
+Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. In windows, differnt versions of a multiarch image could be pulled based on hyperV and process isolation. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
+
+This KEP attempts to goes over the details of the most significant changes needed in containerd, CRI, kubelet and tools like ctr and crictl to support this feature.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, the windows platform matcher looks for an exact OS version match between the host and guest while pulling images. This behavior is not accurate for hyperV isolation as the UVM/sandbox OS version could be different from the host OS version.  
+
+For example, a user might want to pull an image like python (a manifest list) on a WS2022 host with an intent of running it in on a WS2019 hyperV isolated UVM. But unfortunately, the current windows platform matcher behavior only forces the WS2022 version of the python image to be pulled (due to the exact version match between host and guest). If a manifest list does not have a version that matches the host OS version, then the image pull fails.
+Extending image pulls to specify an optional runtime class would greatly help to extend windows platform matcher behavior depending on whether hyperV or process isolation is being used. Additionally, having support to pull images per runtime class might also help projects like supporting different snapshotters with different runtimes as well.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+This KEP describes the detailed design and major API changes needed in kubelet, CRI and containerd to support image pull with optional runtime class option.
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
```

Comment:
snpashotter per runtimeclass is not in scope. It is one of the projects which folks have been talking about addressing for some time now. That needs some additional changes and should be addressed separately. But it will benefit from this work as image will already be addresses by (name, runtimeClass). I mentioned it in the summary/motivation section to only point out snapshotter per runtime class when it happens will benefit from these changes. Should I explicitly call it out as a out of scope/non-goals?

Reply:
@aravindhp  to answer your question "Are changes to tooling like ctr and crictl in or out of scope?" 
No, they are in scope and I do intend to make the necessary changes for these. It is discussed in the later part of this KEP.

Reply:
> Should I explicitly call it out as a out of scope/non-goals?

yes please,  Has there been some discussion or preliminary work done to make sure these changes are compatible?

Reply:
@jsturtevant I made changes to KEP to call it out.
Also, I don't think is any active effort that I know of for supporting snapshotter per runtime class. But we briefly discussed this is the kubelet garbage collection WG with @mikebrow  and @haircommander .